### ISSUE TITLE
ENH: Add `out=` argument to `numpy.astype` function

### ISSUE THREAD

**Issue by vadimkantorov** [numpy/numpy#27141]
### Proposed new feature or change:

Currently it's missing this argument and this is preventing providing a pre-allocated buffer for data type casts. uint8<>float32 / int16<>float32, uint16<>float32 conversions are frequent in audio / image analysis, so having some simple / economic / convenient shortcuts for such conversions would be awesome

Stumbled on it at:
- https://github.com/numpy/numpy/issues/17196

**Comment by seberg** [comment]
Yes, so I am not sure why `asarray` should have it.  There are already at least two ways to achieve the same:
```
arr[...] = values
np.copyto(arr, values)
```
(I know `...` isn't everywhere for some reason, but it correctly indicates N dimensional and thus works for 0-d.)
the only thing both do that `asarray` might reject is broadcasting the values.

**Comment by vadimkantorov** [comment]
I agree that that there are workaround (and thanks for bringing attention to `copyto`!), but many other regular functions already have `out=` argument, so having it supported also for often-allocating dtype conversion `astype()` also makes sense for consistency and makes it less surprising for users

**Comment by jorenham** [comment]
Is there even anything that `asarray` can do, that can't be done in a different way, @seberg?

**Comment by jorenham** [comment]
Perhaps we should even consider making `np.asarray` the new `np.array`, and deprecate the latter. This would be more in line with the array-api, and also with the 13th `import this` rule (I'm Dutch, BTW).
